File: lib/factorix/transfer/downloader.rb
# frozen_string_literal: true

require "pathname"
require "tmpdir"
require "uri"

module Factorix
  module Transfer
    # File downloader with caching and progress tracking
    #
    # Downloads files from HTTPS URLs with automatic caching.
    # Uses file locking to prevent concurrent downloads of the same file.
    # HTTP redirects are handled automatically by the HTTP layer.
    # Reports progress to an optional listener.
    class Downloader
      attr_reader :logger
      attr_reader :cache
      attr_reader :client

      # Dependencies default to the Factorix.app composition root
      def initialize(logger: Factorix.app.logger, cache: Factorix.app.download_cache, client: Factorix.app.download_http_client)
        @logger = logger
        @cache = cache
        @client = client
      end

      # Download a file from the given URL with caching support.
      #
      # If the file exists in cache, it will be copied from cache instead of downloading.
      # If the cached file fails SHA1 verification, it will be invalidated and re-downloaded.
      # If multiple processes attempt to download the same file, only one will download
      # while others wait for the download to complete.
      # HTTP redirects are followed automatically by the HTTP layer.
      #
      # @param url [URI::HTTPS] URL to download from
      # @param output [Pathname] path to save the downloaded file
      # @param expected_sha1 [String, nil] expected SHA1 digest for verification (optional)
      # @param listener [Progress::DownloadHandler, nil] optional progress listener
      # @return [void]
      # @raise [URLError] if the URL is not HTTPS
      # @raise [HTTPClientError] for 4xx HTTP errors
      # @raise [HTTPServerError] for 5xx HTTP errors
      # @raise [DigestMismatchError] if SHA1 verification fails
      def download(url, output, expected_sha1: nil, listener: nil)
        unless url.is_a?(URI::HTTPS)
          logger.error "Invalid URL: must be HTTPS"
          raise URLError, "URL must be HTTPS"
        end

        logger.info("Starting download", output: output.to_s)
        cache_key = strip_query(url)

        case try_cache_hit(cache_key, output, expected_sha1:, listener:)
        when :hit
          return
        when :miss
          logger.debug("Cache miss, downloading", output: output.to_s)
        when :corrupted
          logger.debug("Re-downloading after cache invalidation", output: output.to_s)
        else
          raise RuntimeError, "Unexpected cache state"
        end

        cache.with_lock(cache_key) do
          return if try_cache_hit(cache_key, output, expected_sha1:, listener:) == :hit

          with_temporary_file do |temp_file|
            download_file_with_progress(url, temp_file, listener)
            verify_sha1(temp_file, expected_sha1) if expected_sha1
            cache.store(cache_key, temp_file)
            cache.write_to(cache_key, output)
          end
        end
      end

      # Download file with progress tracking
      #
      # @param url [URI::HTTPS] URL to download from
      # @param output [Pathname] path to save the downloaded file
      # @param listener [Progress::DownloadHandler, nil] optional progress listener
      # @return [void]
      private def download_file_with_progress(url, output, listener)
        current_size = 0

        client.get(url) do |response|
          content_length = response["Content-Length"]
          total_size = content_length ? Integer(content_length, 10) : nil

          listener&.on_started(total: total_size)

          output.open("wb") do |file|
            response.read_body do |chunk|
              file.write(chunk)
              current_size += chunk.bytesize
              listener&.on_progress(current: current_size)
            end
          end
        end

        listener&.on_completed
      end

      # Attempt to retrieve file from cache with SHA1 verification.
      #
      # If the cached file exists but fails SHA1 verification, the cache entry
      # is invalidated and :corrupted is returned to trigger re-download.
      #
      # @param cache_key [String] logical cache key (URL string)
      # @param output [Pathname] path to save the cached file
      # @param expected_sha1 [String, nil] expected SHA1 digest for verification (optional)
      # @param listener [Progress::DownloadHandler, nil] optional progress listener
      # @return [Symbol] :hit if cache hit with valid SHA1, :miss if not cached, :corrupted if SHA1 mismatch
      private def try_cache_hit(cache_key, output, expected_sha1:, listener: nil)
        return :miss unless cache.write_to(cache_key, output)

        logger.info("Cache hit", output: output.to_s)
        verify_sha1(output, expected_sha1) if expected_sha1
        listener&.on_cache_hit(total: cache.size(cache_key))
        :hit
      rescue DigestMismatchError => e
        logger.warn("Cache corrupted, invalidating", output: output.to_s, error: e.message)
        cache.delete(cache_key)
        :corrupted
      end

      # Verify SHA1 digest of a file
      #
      # @param file [Pathname] file to verify
      # @param expected_sha1 [String] expected SHA1 digest
      # @return [void]
      # @raise [DigestMismatchError] if digest does not match
      private def verify_sha1(file, expected_sha1)
        actual_sha1 = Digest(:SHA1).file(file).hexdigest
        return if actual_sha1 == expected_sha1

        logger.error("SHA1 digest mismatch", expected: expected_sha1, actual: actual_sha1)
        raise DigestMismatchError, "SHA1 mismatch: expected #{expected_sha1}, got #{actual_sha1}"
      end

      # Strip query parameters from a URI to create a safe cache key.
      # This prevents sensitive data (e.g., authentication tokens) from being
      # stored in cache metadata.
      #
      # @param uri [URI] the URI to strip
      # @return [String] URI string without query parameters
      private def strip_query(uri) = uri.dup.tap {|u| u.query = nil }.to_s

      # Create a temporary file for downloading, ensuring cleanup after use.
      #
      # @yield [Pathname] the temporary file path
      # @return [void]
      private def with_temporary_file
        dir = Pathname(Dir.mktmpdir("factorix"))
        temp_file = dir.join("download")
        temp_file.binwrite("")
        yield temp_file
      ensure
        temp_file&.unlink if temp_file&.exist?
        dir&.rmdir if dir&.exist?
      end
    end
  end
end

File: internal/transfer/downloader.go
// Package transfer downloads and uploads files with caching and progress
// reporting.
package transfer

import (
	"context"
	"crypto/sha1"
	"encoding/hex"
	"fmt"
	"io"
	"log/slog"
	"net/url"
	"os"
	"path/filepath"

	"github.com/sakuro/factorix/internal/cache"
	"github.com/sakuro/factorix/internal/httpx"
	"github.com/sakuro/factorix/internal/progress"
)

// Downloader downloads files with caching. Concurrent downloads of the same
// URL are serialized by the cache lock, so only one process fetches while
// the others wait and then hit the cache.
type Downloader struct {
	Cache  cache.Cache
	Client *httpx.Client // retry only; caching is handled here, not by CacheTransport
	Logger *slog.Logger
}

// NewDownloader builds a downloader.
func NewDownloader(c cache.Cache, client *httpx.Client, logger *slog.Logger) *Downloader {
	if logger == nil {
		logger = slog.New(slog.NewTextHandler(io.Discard, nil))
	}
	return &Downloader{Cache: c, Client: client, Logger: logger}
}

// DownloadOptions modify a download.
type DownloadOptions struct {
	// ExpectedSHA1 enables digest verification of both cached and freshly
	// downloaded content.
	ExpectedSHA1 string
	Listener     progress.Listener
}

// Download fetches rawURL to the output path, serving from the cache when
// possible. A cached file failing SHA1 verification is invalidated and
// re-downloaded.
func (d *Downloader) Download(ctx context.Context, rawURL, output string, opts DownloadOptions) error {
	cacheKey, err := stripQuery(rawURL)
	if err != nil {
		return err
	}
	d.Logger.Info("Starting download", "output", output)

	hit, err := d.tryCacheHit(ctx, cacheKey, output, opts)
	if err != nil {
		return err
	}
	if hit {
		return nil
	}
	d.Logger.Debug("Cache miss, downloading", "output", output)

	return d.Cache.WithLock(ctx, cacheKey, func() error {
		// Another process may have completed the download while we waited.
		hit, err := d.tryCacheHit(ctx, cacheKey, output, opts)
		if err != nil || hit {
			return err
		}

		tempDir, err := os.MkdirTemp("", "factorix-download")
		if err != nil {
			return err
		}
		defer os.RemoveAll(tempDir)
		tempFile := filepath.Join(tempDir, "download")

		if err := d.downloadWithProgress(ctx, rawURL, tempFile, opts.Listener); err != nil {
			return err
		}
		if opts.ExpectedSHA1 != "" {
			if err := verifySHA1(tempFile, opts.ExpectedSHA1); err != nil {
				return err
			}
		}
		if _, err := d.Cache.Store(ctx, cacheKey, tempFile); err != nil {
			return err
		}
		// Copy from the temp file, not the cache, so a store skipped by a
		// size limit still produces the output.
		return copyFile(tempFile, output)
	})
}

// tryCacheHit writes cached content to output. On a SHA1 mismatch the entry
// is invalidated and (false, nil) is returned to trigger a re-download.
func (d *Downloader) tryCacheHit(ctx context.Context, cacheKey, output string, opts DownloadOptions) (bool, error) {
	found, err := d.Cache.WriteTo(ctx, cacheKey, output)
	if err != nil || !found {
		return false, err
	}
	d.Logger.Info("Cache hit", "output", output)

	if opts.ExpectedSHA1 != "" {
		if err := verifySHA1(output, opts.ExpectedSHA1); err != nil {
			d.Logger.Warn("Cache corrupted, invalidating", "output", output, "error", err)
			if _, deleteErr := d.Cache.Delete(ctx, cacheKey); deleteErr != nil {
				return false, deleteErr
			}
			return false, nil
		}
	}

	if info, err := os.Stat(output); err == nil {
		progress.Start(opts.Listener, info.Size())
		progress.Update(opts.Listener, info.Size())
		progress.Finish(opts.Listener)
	}
	return true, nil
}

func (d *Downloader) downloadWithProgress(ctx context.Context, rawURL, output string, listener progress.Listener) error {
	resp, err := d.Client.GetStream(ctx, rawURL)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	progress.Start(listener, resp.ContentLength) // -1 when unknown

	file, err := os.Create(output)
	if err != nil {
		return err
	}
	defer file.Close()

	counter := &countingWriter{listener: listener}
	if _, err := io.Copy(io.MultiWriter(file, counter), resp.Body); err != nil {
		return err
	}
	if err := file.Close(); err != nil {
		return err
	}
	progress.Finish(listener)
	return nil
}

// countingWriter forwards the cumulative byte count to the listener.
type countingWriter struct {
	listener progress.Listener
	current  int64
}

func (w *countingWriter) Write(p []byte) (int, error) {
	w.current += int64(len(p))
	progress.Update(w.listener, w.current)
	return len(p), nil
}

func verifySHA1(path, expected string) error {
	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()

	h := sha1.New()
	if _, err := io.Copy(h, f); err != nil {
		return err
	}
	actual := hex.EncodeToString(h.Sum(nil))
	if actual != expected {
		return fmt.Errorf("%w: expected %s, got %s", ErrDigestMismatch, expected, actual)
	}
	return nil
}

// stripQuery removes the query from the URL so credentials (username/token)
// never end up in cache keys or metadata.
func stripQuery(rawURL string) (string, error) {
	u, err := url.Parse(rawURL)
	if err != nil {
		return "", err
	}
	u.RawQuery = ""
	return u.String(), nil
}

func copyFile(src, dst string) error {
	data, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	return os.WriteFile(dst, data, 0o644)
}
